File: plugins/plugin-discord/typescript/actions/createPoll.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  composePromptFromState,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  type Memory,
  ModelType,
  parseJSONObjectFromText,
  type State,
} from "@elizaos/core";
import type { TextChannel } from "discord.js";
import { DISCORD_SERVICE_NAME } from "../constants";
// Import generated prompts
import { createPollTemplate } from "../generated/prompts/typescript/prompts.js";
import type { DiscordService } from "../service";

const getPollInfo = async (
  runtime: IAgentRuntime,
  _message: Memory,
  state: State
): Promise<{
  question: string;
  options: string[];
  useEmojis: boolean;
} | null> => {
  const prompt = composePromptFromState({
    state,
    template: createPollTemplate,
  });

  for (let i = 0; i < 3; i++) {
    const response = await runtime.useModel(ModelType.TEXT_SMALL, {
      prompt,
    });

    const parsedResponse = parseJSONObjectFromText(response);
    if (
      parsedResponse?.question &&
      Array.isArray(parsedResponse.options) &&
      parsedResponse.options.length >= 2
    ) {
      return {
        question: String(parsedResponse.question),
        options: parsedResponse.options.slice(0, 10).map(String), // Max 10 options
        useEmojis: parsedResponse.useEmojis !== false, // Default to true
      };
    }
  }
  return null;
};

// Number emojis for poll options
const numberEmojis = ["1️⃣", "2️⃣", "3️⃣", "4️⃣", "5️⃣", "6️⃣", "7️⃣", "8️⃣", "9️⃣", "🔟"];
const letterEmojis = ["🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬", "🇭", "🇮", "🇯"];
const yesNoEmojis = ["✅", "❌"];

export const createPoll: Action = {
  name: "CREATE_POLL",
  similes: [
    "CREATE_POLL",
    "MAKE_POLL",
    "START_POLL",
    "CREATE_VOTE",
    "MAKE_VOTE",
    "START_VOTE",
    "CREATE_SURVEY",
  ],
  description: "Create a poll in Discord with emoji reactions for voting.",
  validate: async (_runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    return message.content.source === "discord";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const discordService = runtime.getService(DISCORD_SERVICE_NAME) as DiscordService;

    if (!discordService || !discordService.client) {
      if (callback) {
        await callback?.({
          text: "Discord service is not available.",
          source: "discord",
        });
      }
      return { success: false, error: "Discord service is not available" };
    }

    if (!state) {
      if (callback) {
        await callback?.({
          text: "State is not available.",
          source: "discord",
        });
      }
      return { success: false, error: "State is not available" };
    }

    const pollInfo = await getPollInfo(runtime, message, state);
    if (!pollInfo) {
      if (callback) {
        await callback?.({
          text: "I couldn't understand the poll details. Please specify a question and at least 2 options.",
          source: "discord",
        });
      }
      return { success: false, error: "Could not parse poll details" };
    }

    try {
      const stateData = state.data;
      const room = stateData?.room || (await runtime.getRoom(message.roomId));
      if (!room || !room.channelId) {
        if (callback) {
          await callback?.({
            text: "I couldn't determine the current channel.",
            source: "discord",
          });
        }
        return { success: false, error: "Could not determine current channel" };
      }

      const channel = await discordService.client.channels.fetch(room.channelId);
      if (!channel || !channel.isTextBased()) {
        if (callback) {
          await callback?.({
            text: "I can only create polls in text channels.",
            source: "discord",
          });
        }
        return { success: false, error: "Channel is not a text channel" };
      }

      const textChannel = channel as TextChannel;

      // Determine which emojis to use
      let emojis: string[];
      if (
        pollInfo.options.length === 2 &&
        pollInfo.options.some((opt) => opt.toLowerCase().includes("yes")) &&
        pollInfo.options.some((opt) => opt.toLowerCase().includes("no"))
      ) {
        emojis = yesNoEmojis;
      } else if (pollInfo.useEmojis) {
        emojis = numberEmojis.slice(0, pollInfo.options.length);
      } else {
        emojis = letterEmojis.slice(0, pollInfo.options.length);
      }

      // Format the poll message
      const pollMessage = [
        `📊 **POLL: ${pollInfo.question}**`,
        "",
        ...pollInfo.options.map((option, index) => `${emojis[index]} ${option}`),
        "",
        "_React to vote!_",
      ].join("\n");

      // Send the poll message
      const sentMessage = await textChannel.send(pollMessage);

      // Add reactions
      for (let i = 0; i < pollInfo.options.length; i++) {
        try {
          await sentMessage.react(emojis[i]);
          // Small delay to avoid rate limits
          await new Promise((resolve) => setTimeout(resolve, 250));
        } catch (error) {
          runtime.logger.error(
            {
              src: "plugin:discord:action:create-poll",
              agentId: runtime.agentId,
              emoji: emojis[i],
              error: error instanceof Error ? error.message : String(error),
            },
            "Failed to add reaction"
          );
        }
      }

      const response: Content = {
        text: `I've created a poll with ${pollInfo.options.length} options. Users can vote by clicking the reaction emojis!`,
        source: message.content.source,
      };

      if (callback) {
        await callback?.(response);
      }
      return { success: true, text: response.text };
    } catch (error) {
      runtime.logger.error(
        {
          src: "plugin:discord:action:create-poll",
          agentId: runtime.agentId,
          error: error instanceof Error ? error.message : String(error),
        },
        "Error creating poll"
      );
      if (callback) {
        await callback?.({
          text: "I encountered an error while creating the poll. Please make sure I have permission to send messages and add reactions.",
          source: "discord",
        });
      }
      return { success: false, error: error instanceof Error ? error.message : String(error) };
    }
  },
  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "create a poll: What game should we play tonight? Options: Minecraft, Fortnite, Valorant",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "I'll create a poll for game selection with those options.",
          actions: ["CREATE_POLL"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "make a vote: Should we have the meeting at 3pm? Yes/No",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Creating a yes/no poll about the meeting time.",
          actions: ["CREATE_POLL"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "start a poll asking what day works best for everyone: Monday, Tuesday, Wednesday, Thursday, Friday",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "I'll create a poll to find the best day for everyone.",
          actions: ["CREATE_POLL"],
        },
      },
    ],
  ] as ActionExample[][],
};

export default createPoll;

File: plugins/plugin-discord/rust/src/actions/create_poll.rs
//! Create poll action

use async_trait::async_trait;

use super::{ActionContext, ActionResult, DiscordAction};
use crate::error::Result;
use crate::types::Snowflake;
use crate::DiscordService;

/// Emoji sets for polls
const NUMBER_EMOJIS: [&str; 10] = ["1️⃣", "2️⃣", "3️⃣", "4️⃣", "5️⃣", "6️⃣", "7️⃣", "8️⃣", "9️⃣", "🔟"];
const LETTER_EMOJIS: [&str; 10] = ["🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬", "🇭", "🇮", "🇯"];
const YES_NO_EMOJIS: [&str; 2] = ["✅", "❌"];

/// Action to create a poll in Discord with emoji reactions for voting
pub struct CreatePollAction;

#[async_trait]
impl DiscordAction for CreatePollAction {
    fn name(&self) -> &str {
        "CREATE_POLL"
    }

    fn description(&self) -> &str {
        "Create a poll in Discord with emoji reactions for voting."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "MAKE_POLL",
            "START_POLL",
            "CREATE_VOTE",
            "MAKE_VOTE",
            "START_VOTE",
            "CREATE_SURVEY",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context
            .message
            .get("source")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        Ok(source == "discord")
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &DiscordService,
    ) -> Result<ActionResult> {
        let channel_id = Snowflake::new(context.channel_id.clone())?;

        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        // Parse poll info from the message
        let poll_info = service.parse_poll_info(text).await?;

        let poll_info = match poll_info {
            Some(info) => info,
            None => {
                return Ok(ActionResult::failure(
                    "I couldn't understand the poll details. \
                    Please specify a question and at least 2 options.",
                ))
            }
        };

        let question = poll_info
            .get("question")
            .and_then(|q| q.as_str())
            .unwrap_or("");
        let options: Vec<String> = poll_info
            .get("options")
            .and_then(|o| o.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .take(10)
                    .collect()
            })
            .unwrap_or_default();
        let use_emojis = poll_info
            .get("use_emojis")
            .and_then(|u| u.as_bool())
            .unwrap_or(true);

        if options.len() < 2 {
            return Ok(ActionResult::failure("A poll needs at least 2 options."));
        }

        // Determine which emojis to use
        let emojis: Vec<&str> = if options.len() == 2
            && options.iter().any(|o| o.to_lowercase().contains("yes"))
            && options.iter().any(|o| o.to_lowercase().contains("no"))
        {
            YES_NO_EMOJIS.to_vec()
        } else if use_emojis {
            NUMBER_EMOJIS[..options.len()].to_vec()
        } else {
            LETTER_EMOJIS[..options.len()].to_vec()
        };

        // Format the poll message
        let mut poll_lines = vec![format!("📊 **POLL: {}**", question), String::new()];
        for (i, option) in options.iter().enumerate() {
            poll_lines.push(format!("{} {}", emojis[i], option));
        }
        poll_lines.push(String::new());
        poll_lines.push("_React to vote!_".to_string());
        let poll_message = poll_lines.join("\n");

        // Send the poll
        let message_id = service
            .send_poll(&channel_id, &poll_message, &emojis[..options.len()])
            .await?;

        match message_id {
            Some(id) => Ok(ActionResult::success_with_data(
                format!(
                    "I've created a poll with {} options. \
                    Users can vote by clicking the reaction emojis!",
                    options.len()
                ),
                serde_json::json!({
                    "message_id": id,
                    "question": question,
                    "options": options,
                }),
            )),
            None => Ok(ActionResult::failure(
                "Failed to create the poll. Please check my permissions.",
            )),
        }
    }
}
